text/0000-otlp-trace-data-format.md

```diff
@@ -0,0 +1,376 @@
+// which is used by [gRPC](https://github.com/grpc).
+message Status {
+  // The status code. This is optional field. It is safe to assume 0 (OK)
+  // when not set.
+  int32 code = 1;
+
+  // A developer-facing error message, which should be in English.
+  string message = 2;
+}
+```
+
+### AttributeKeyValue
+
+```
+message AttributeKeyValue {
+  enum ValueType {
+    STRING  = 0;
+    BOOL    = 1;
+    INT64   = 2;
+    FLOAT64 = 3;
+    BINARY  = 4;
+  };
+
+  string key = 1;
+  // The type of the value.
+  ValueType type = 2;
+  // A string up to 256 bytes long.
+  string string_value = 3;
+  // A 64-bit signed integer.
+  int64 int_value = 4;
+  // A Boolean value represented by `true` or `false`.
+  bool bool_value = 5;
+  // A double value.
+  double double_value = 6;
+  // A binary value of bytes.
+  bytes binary_value = 7;
+}
+
+```
+
+## Trade-offs and mitigations
+
+Timestamps were changed from google.protobuf.Timestamp to a int64 representation in Unix epoch nanoseconds. This change reduces the type-safety but benchmarks show that for small spans there is 15-20% encoding/decoding CPU speed gain. This is the right trade-off to make because encoding/decoding CPU consumption tends to dominate many workloads (particularly in OpenTelemetry Service).
+
+## Prior art and alternatives
+
+OpenCensus and Jaeger protocol buffer data schemas were used as the inspiration for this specification. OpenCensus was the starting point, Jaeger provided performance improvement ideas.
+
+## Open questions
+
+A follow up RFC is required to define the data format for metrics.
+
+## Appendix A - Benchmarking
+
+The following shows [benchmarking](https://github.com/tigrannajaryan/exp-otelproto/) using various schemas.
+
+Legend:
+- OpenCensus    - OpenCensus protocol schema.
+- OTLP/AttrMap  - OTLP schema using map for attributes.
+- OTLP/AttrList - OTLP schema using list of key/values for attributes and with reduced nesting for values.
+- OTLP/AttrList/TimeWrapped - Same as OTLP/AttrList, except using google.protobuf.Timestamp instead of int64 for timestamps.
+
+```
+BenchmarkEncode/OpenCensus-8         	      10	 622843375 ns/op
+BenchmarkEncode/OTLP/AttrMap-8       	      10	 528652501 ns/op
+BenchmarkEncode/OTLP/AttrList-8      	      50	 132795894 ns/op
+BenchmarkEncode/OTLP/AttrList/TimeWrapped-8         	      50	 154429659 ns/op
+BenchmarkDecode/OpenCensus-8                        	      10	 656730934 ns/op
+BenchmarkDecode/OTLP/AttrMap-8                      	      10	 567621380 ns/op
+BenchmarkDecode/OTLP/AttrList-8                     	      30	 240150634 ns/op
+BenchmarkDecode/OTLP/AttrList/TimeWrapped-8         	      30	 293842943 ns/op
+```
+
+Benchmarks show OTLP/AttrList is 4.7 times faster than OpenCensus in encoding and 2.7 times faster in decoding.
+
+Using google.protobuf.Timestamp results in 1.16 times slower encoding and 1.22 times slower decoding.
```

Comment:
You need to make sure you specify the language. In C++ this will not be true because probably most of the cost comes from allocating extra objects.

Also what data do you send? How many attributes/events?

Reply:
You are right.
The benchmark uses batches of 100 spans, each span containing 3 attributes, the total uncompressed, encoded size of each batch is around 20KBytes.
I will add this info to the doc.

Reply:
I added the info to the doc.